test/framework/delete_cluster.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	dockerv1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DeleteCluster deletes the cluster
+// Assertions:
+//   * Deletes Machines
+//   * Deletes MachineSets
+//   * Deletes MachineDeployments
+//   * Deletes KubeadmConfigs
+//   * Deletes InfraCluster
+//   * Deletes InfraMachines
+//   * Deletes InfraMachineTemplates
+//   * Deletes Secrets
+func DeleteCluster(input *MultiNodeControlplaneClusterInput) {
```

Comment:
I think we could wrap this up in Cleanup. That way every test would also make assertions that everything is properly cleaned up and we wouldn't be adding an extra `n` minutes to the run. I think Ginkgo is sufficiently good at printing logs such that pinpointing delete as the culprit should be fairly easy.

Reply:
Interestingly, I started this work by adding it into `CleanUp` and then decided it would be _"faster"_ to just create a separate test `DeleteCluster` because currently `CleanUp` takes a specific `CleanUpInput` struct. 

I like @detiber's idea about putting a cleanup method on the MultiNode or OneNode's objects so that test writers can just do something like 
```
framework.OneNodeCluster(input)
input.CleanUp()
```
Unless I misunderstood your comment @detiber.

I can do the refactoring as part of this PR but have a separate commit so it is easier to review. WDYT?

Reply:
I was also hoping to combine the `OneNodeClusterInput` and `MultiNodeControlplaneClusterInput` since they are essentially the same except for the `Node` vs `[]Node` fields. 
If these are combined into a single object for now then we could just have a single CleanUp function for the cluster input. Unless there is some other reason we have the separation?

Reply:
that sounds like a great refactor to me

Reply:
> I can do the refactoring as part of this PR but have a separate commit so it is easier to review. WDYT?

+1, thanks

Reply:
All of that sounds good to me :)